server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -0,0 +1,164 @@
+
+package org.elasticsearch.synonyms;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.bulk.BulkRequestBuilder;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.action.synonyms.PutSynonymsAction;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.client.internal.OriginSettingClient;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.reindex.DeleteByQueryAction;
+import org.elasticsearch.index.reindex.DeleteByQueryRequest;
+import org.elasticsearch.indices.SystemIndexDescriptor;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+
+import static org.elasticsearch.index.mapper.MapperService.SINGLE_MAPPING_NAME;
+import static org.elasticsearch.xcontent.XContentFactory.jsonBuilder;
+
+public class SynonymsManagementAPIService {
+    public static final String SYNONYMS_INDEX = ".synonyms";
+    public static final String SYNONYMS_ORIGIN = "synonyms";
+
+    public static final String SYNONYMS_FEATURE_NAME = "synonyms";
+    public static final String SYNONYM_SET_FIELD = "synonym_set";
+    public static final String SYNONYM_FIELD = "synonym";
+
+    private final Client client;
+
+    public static final SystemIndexDescriptor SYNONYMS_DESCRIPTOR = SystemIndexDescriptor.builder()
+        .setIndexPattern(SYNONYMS_INDEX + "*")
+        .setDescription("Synonyms index for synonyms managed through APIs")
+        .setPrimaryIndex(SYNONYMS_INDEX)
+        .setMappings(mappings())
+        .setSettings(settings())
+        .setVersionMetaKey("version")
+        .setOrigin(SYNONYMS_ORIGIN)
+        .build();
+
+    public SynonymsManagementAPIService(Client client) {
+        // TODO Should we set an OriginSettingClient? We would need to check the origin at AuthorizationUtils if we set an
+        this.client = new OriginSettingClient(client, SYNONYMS_ORIGIN);
+    }
+
+    private static XContentBuilder mappings() {
+        try {
+            XContentBuilder builder = jsonBuilder();
+            builder.startObject();
+            {
+                builder.startObject(SINGLE_MAPPING_NAME);
+                {
+                    builder.startObject("_meta");
+                    {
+                        builder.field("version", Version.CURRENT.toString());
+                    }
+                    builder.endObject();
+                    builder.field("dynamic", "strict");
+                    builder.startObject("properties");
+                    {
+                        builder.startObject("synonym");
```

Comment:
I wanted to prevent confusion so `synonyms` would refer to the topmost synonym set, and `synonym` to the individual rule:

```
{
    "synonyms": [
        {
            "id": "32rfs",
            "synonym": "hello, hi"
        },
        {          
            "synonym": "bye, goodbye"
        },
        {          
            "synonym": "test => check"
        }
    ]
}
```

Maybe we could use synonym_set for the topmost synonym element, and synonyms for the inner ones:

```
{
    "synonym_set": [
        {
            "id": "32rfs",
            "synonyms": "hello, hi"
        },
        {          
            "synonyms": "bye, goodbye"
        },
        {          
            "synonyms": "test => check"
        }
    ]
}
```

WDYT?

Reply:
We got an agreement to use:

```json
{
    "synonyms_set": [
        {
            "id": "32rfs",
            "synonyms": "hello, hi"
        },
        {          
            "synonyms": "bye, goodbye"
        },
        {          
            "synonyms": "test => check"
        }
    ]
}
```

I'll be updating the code with this 👍